docs/design/sql/14-automatic-query-resubmission.md

```diff
@@ -0,0 +1,157 @@
+# Automatic query resubmission
+
+|||
+|---|------------------- https://hazelcast.atlassian.net/browse/HZ-1098 https://hazelcast.atlassian.net/browse/HZ-1098 |
+|Related Github issues|https://github.com/hazelcast/hazelcast/pull/21635|
+|Document Status / Completeness| Complete                                          |
+|Requirement owner| Sandeep Akhouri                                              |
+|TDD Author|Viliam Ďurina|
+|Developer(s)| Krzysztof Ślusarski |
+|Quality Engineer| TODO                             |
+|Support Engineer| TODO                             |
+|Technical Reviewers| TODO                          |
+|Simulator or Soak Test PR(s) | TODO            |
+
+### Background
+#### Description
+
+A customer perceives that AP is NOT implemented correctly in Hazelcast. If the
+member hosting the primary partition is NOT available, their expectation is that
+data should be transparently served from the backup. When the cluster is
+changing topology, e.g., cluster member that contains the primary data is
+killed, the client throws errors.
+
+Automatic retries, with an upper bound on the number of retries, are implemented
+with many IMap operations on the client. However, with SQL, the client cannot do
+this, because it doesn't understand the query, and it doesn't know if the query
+is side-effect-free and it's safe to be retried. Secondly, the client might
+already have received some result rows, and if it retries the query, those rows
+will be received again.
+
+#### Scope
+
+The scope is to add support for automatic SQL retries in Java client. Other
+clients might follow, but it's not in the scope of the initial PR, as the
+requirement is currently only for Java client.
+
+There will be no retry mechanism for queries submitted through member instances.
```

Comment:
Good question actually. Perhaps it would be easy to implement for member instances too, except for not retrying in case the local member is the cause.

Reply:
As we discussed, this is not supported because on member instances the concept client session is unclear. In clients, the client session is equal to a single connection. But on member instances, most likely, parallel queries are run. From this follows that we'd need the per-query configuration, which we don't want to add for reasons mentioned in the TDD.

However, if there's a need, we can add this behavior in the future.